### ISSUE TITLE
`UnreachableCode` false positive for `?: return` when type is in another module

### ISSUE THREAD

**Issue by kylannjohnson** [detekt/detekt#5186]
## Expected Behavior
If a nullable parameter is passed into a method and the method checks it for null and returns immediately, the code in the method after the check should not be marked as unreachable

## Observed Behavior
Consider this code where `Foo` is defined in separate gradle module

```kotlin
fun fooCheck(foo: Foo?) {
    val nonNullFoo = foo?.aBar ?: return

    println("$nonNullFoo was not null")
}

// Foo is:    data class Foo(val aBar: String?)
```

When you run `detektMain`, it returns

```
UnreachableCode - [This expression is unreachable code which should either be used or removed.] at /Foo.kt:6:5
UnreachableCode - [This expression is unreachable code which should either be used or removed.] at /Foo.kt:8:5
```

However, if you change it to `LocalFoo`, then `detektMain` is successful

```kotlin
fun fooCheck(foo: LocalFoo?) {
    val nonNullFoo = foo?.aBar ?: return

    println("$nonNullFoo was not null")
}

// Same as Foo above
data class LocalFoo(val aBar: String?) 
```

## Your Environment

* Version of detekt used: project is on 1.20.0 but 1.21.0 is the same
* Version of Gradle used (if applicable): 7.5
* Operating System and version: macOs 12.1

**Comment by BraisGabin** [comment]
Does `detektMain` complains with a message like:

> `The BindingContext was created with N issues`

In general detekt shouldn't fail with classes that are defined in different modules. But if those classes are injected in the classpath in a "funky" way it fails. That usually happes if the classes are related with Android.

---

Anyway I could reproducte your issue with this Unit test:

```kotlin
@Test
fun `does not report unreachable code with unknown types`() {
    val code = """
        fun fooCheck(foo: Foo?) {
            foo?.aBar ?: return
        }
    """
    assertThat(subject.lintWithContext(env, code)).isEmpty()
}
```

The problem is that this rule just borrow the information from the `bindingContext`. And if it wasn't generated properly there's really little we can do here.

The issues regarding the `BindingContext` generation are known and we are trying to fix them. But they aren't easy.

**Comment by kylannjohnson** [comment]
Yes, it does.
> `The BindingContext was created with 307 issues. Run detekt with --debug to see the error messages.`

If it helps, both modules (the violating module and the model defining module) are Kotlin modules, but there is Android in the overall project.  Thanks for working on the issue.

**Comment by matejdro** [comment]
Can we get this reopened? It still appears to be a problem in 1.22.0

**Comment by atulgpt** [comment]
> ```kotlin
> @Test
> fun `does not report unreachable code with unknown types`() {
>     val code = """
>         fun fooCheck(foo: Foo?) {
>             foo?.aBar ?: return
>         }
>     """
>     assertThat(subject.lintWithContext(env, code)).isEmpty()
> }
> ```

Hi, @BraisGabin this ut is not failing. Could this be fixed by https://github.com/JetBrains/kotlin/commit/e2b7aba086238af5c77618b8caadd6a8eb82e6ce ?

**Comment by BraisGabin** [comment]
Closing then.

**Comment by scherrsasrf** [comment]
@BraisGabin this issue has popped up in our project with an upgrade to kotlin 2.2.0. Seems whatever fixed it is now broken again...
Setup is the same as above, it only occurs when the referenced type is in another gradle module.

**Comment by karoljur** [comment]
This issue also popped up for me after an upgrade of android gradle project - kotlin dependency to version 2.2.0. 
It seems like issue always relate to return statement inside of collection mappers, i.e.            
```
anything.mapNotNull {
      xyz ?: return@mapNotNull null
}
```

**Comment by psemenov-tempo** [comment]
Hello, the same, issue reproducible with Kotlin 2.2.0

**Comment by marschwar** [comment]
I am unable to reproduce the issue. I have created a project with 2 modules `app` and `utils`. 

In `utils` there is a data class `Model`. `app` depends on `utils`and contains a class `Consumer` like this.

```kotlin
package dev.detekt.app

import dev.detekt.utils.Model

class Consumer {
    fun printName(model: Model?) {
        val name = model?.name ?: return

        println("Name: $name")
    }
}
```

I tried with kotlin 2.1.21 and 2.2.0. In both cases no issues were found.

Are you able to share a reproducer?

**Comment by meowora** [comment]
This issue also occurred for me on kotlin 2.2.0

**Comment by meowora** [comment]
> Are you able to share a reproducer?

While not directed at me i still can share a repo that has this issue.
https://github.com/meowdding/SkyOcean started to have these problems after we bumped kotlin from 2.1.20 to 2.2.0

(one) [line that fails](https://github.com/meowdding/SkyOcean/blob/05c8bbe5c4f5e8e814b3eb6bbe7fb329fdc60f1d/src/common/main/kotlin/me/owdding/skyocean/features/misc/PetCandy.kt#L34)
[full build log](https://github.com/meowdding/SkyOcean/actions/runs/16995389769)

**Comment by marschwar** [comment]
Thank you for providing the repo. With the current `main` branch and the `UnreachableCode` rule re-enabled, I ran the `detektMain` task without any issues. To double check, I added some unreachable code and it was found the issue.

The commit `05c8bbe` you referenced does indeed have the issue. Changing the detekt version from 1.23.7 -> 1.23.8 did fix it.

@psemenov-tempo, @karoljur, @scherrsasrf Which version of detekt are you using?

**Comment by karoljur** [comment]
When I reproduced this issue, our app was for sure on detekt v 1.23.8. Sorry, I can't share real code.
Your reproduction example is incorrect - code which Marschawar presented just do return if some field is null (val name = model?.name ?: return). Just like I wrote before - issue exist in scenario, where we iterate over **COLLECTION** using mapNotNull, then returning if element was null from INSIDE OF MAPNOTNULL.
I.e. 
```
class Consumer {
    private fun validateNames(models: List<Model>) {
        models.mapNotNull {
              val name = it?.name ?: return@mapNotNull null      (this line will be marked as UnreachableCode)
              val whatever = it?.whatever ?: return@mapNotNull null

              println("Name: $name")
    }
}
```

**Comment by marschwar** [comment]
Still no luck @karoljur . Please have a look at https://github.com/marschwar/detekt-reproducer-5186 where I tried to reproduce the issue. Using detekt 1.23.7 will produce UnreachableCode violations, while 1.23.8 does not. Let me know how to change the code.

**Comment by yogurtearl** [comment]
Not stale?
